In [1]:
# Paso 1: Carga de Datos
import pandas as pd

train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')


In [2]:
# Paso 2: Split de datos
from sklearn.model_selection import train_test_split

train_data, test_data = train_test_split(train, test_size=0.2, random_state=42)

In [3]:
# Paso 3: EDA y Profiling de los datos
from pandas_profiling import ProfileReport


C:\Users\gehov\AppData\Local\Temp\ipykernel_11516\91327901.py:2: DeprecationWarning: `import pandas_profiling` is going to be deprecated by April 1st. Please use `import ydata_profiling` instead.
  from pandas_profiling import ProfileReport


In [4]:
# Paso 4: Descripción de los warnings generados en el EDA
# Analizar los resultados del profiling y abordar cualquier problema identificado.

# Verificar la presencia de valores nulos
missing_values = train_data.isnull().sum()
print("Valores nulos por columna:")
print(missing_values[missing_values > 0])

# Verificar duplicados
duplicates = train_data.duplicated()
print("\nNúmero de duplicados:", duplicates.sum())

# Otras comprobaciones según tus necesidades

# ...


Valores nulos por columna:
LotFrontage      217
Alley           1094
MasVnrType         6
MasVnrArea         6
BsmtQual          28
BsmtCond          28
BsmtExposure      28
BsmtFinType1      28
BsmtFinType2      28
Electrical         1
FireplaceQu      547
GarageType        64
GarageYrBlt       64
GarageFinish      64
GarageQual        64
GarageCond        64
PoolQC          1162
Fence            935
MiscFeature     1122
dtype: int64

Número de duplicados: 0


In [5]:
# Paso 5: Setup - Configuración de PyCaret
from pycaret.regression import *

reg = setup(data=train_data, 
            target='SalePrice',
            numeric_imputation='mean',
            categorical_features=['MSZoning','Exterior1st','Exterior2nd','KitchenQual','Functional','SaleType',
                                  'Street','LotShape','LandContour','LotConfig','LandSlope','Neighborhood',   
                                  'Condition1','Condition2','BldgType','HouseStyle','RoofStyle','RoofMatl',    
                                  'MasVnrType','ExterQual','ExterCond','Foundation','BsmtQual','BsmtCond',   
                                  'BsmtExposure','BsmtFinType1','BsmtFinType2','Heating','HeatingQC','CentralAir',   
                                  'Electrical','GarageType','GarageFinish','GarageQual','GarageCond','PavedDrive',
                                  'SaleCondition'],
            ignore_features=['Alley','PoolQC','MiscFeature','Fence','FireplaceQu','Utilities'],
            feature_selection=True,
            pca=True,
            remove_multicollinearity=True,
            remove_outliers=True,
            normalize=True,
            verbose=True)


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.010118 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 64704
[LightGBM] [Info] Number of data points in the train set: 776, number of used features: 254
[LightGBM] [Info] Start training from score 178918.916237
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive 

In [6]:
# Paso 6: Entrenamiento y selección de modelos automática
compare_models(sort='RMSE') 


LGBMRegressor(n_jobs=-1, random_state=5693)

In [7]:
# Seleccionar los tres mejores modelos
top3_models = compare_models(n_select=3, sort='RMSE')

In [8]:
top3_models

[LGBMRegressor(n_jobs=-1, random_state=5693),
 ExtraTreesRegressor(n_jobs=-1, random_state=5693),
 RandomForestRegressor(n_jobs=-1, random_state=5693)]

In [28]:
# Selecciona los tres mejores modelos y entrena
top3_models = compare_models(n_select=3, sort='RMSE', verbose=False)

# Muestra la gráfica de importancia de características para cada modelo
for model in top3_models:
    plot_model(model, plot='feature', save=True)



In [9]:
# Paso 7: Optimización de hiper-parámetros
tuned_models = [tune_model(model) for model in top3_models]

Fitting 10 folds for each of 10 candidates, totalling 100 fits
Original model was better than the tuned model, hence it will be returned. NOTE: The display metrics are for the tuned model (not the original one).


Fitting 10 folds for each of 10 candidates, totalling 100 fits


Fitting 10 folds for each of 10 candidates, totalling 100 fits


In [10]:
# Paso 8: Selección del modelo final
best_model = automl(optimize='rmse')

In [11]:
# Paso 9: Almacenamiento del pipeline en el disco duro
save_model(best_model, 'house_price_model')

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=Memory(location=None),
          steps=[('numerical_imputer',
                  TransformerWrapper(include=['Id', 'MSSubClass', 'LotFrontage',
                                              'LotArea', 'OverallQual',
                                              'OverallCond', 'YearBuilt',
                                              'YearRemodAdd', 'MasVnrArea',
                                              'BsmtFinSF1', 'BsmtFinSF2',
                                              'BsmtUnfSF', 'TotalBsmtSF',
                                              '1stFlrSF', '2ndFlrSF',
                                              'LowQualFinSF', 'GrLivArea',
                                              'BsmtFullBath', 'BsmtHalfBath',
                                              'FullBath', 'Hal...
                  TransformerWrapper(transformer=RemoveOutliers(random_state=5693))),
                 ('normalize', TransformerWrapper(transformer=StandardScaler())),
             

In [12]:
# Paso 10: Predicciones
loaded_model = load_model('house_price_model')
predictions = predict_model(loaded_model, data=test_data)

Transformation Pipeline and Model Successfully Loaded


In [21]:
predictions

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
892    893          20       RL         70.0     8414   Pave   NaN      Reg   
1105  1106          60       RL         98.0    12256   Pave   NaN      IR1   
413    414          30       RM         56.0     8960   Pave  Grvl      Reg   
522    523          50       RM         50.0     5000   Pave   NaN      Reg   
1036  1037          20       RL         89.0    12898   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
479    480          30       RM         50.0     5925   Pave   NaN      Reg   
1361  1362          20       RL        124.0    16158   Pave   NaN      IR1   
802    803          60       RL         63.0     8199   Pave   NaN      Reg   
651    652          70       RL         60.0     9084   Pave   NaN      Reg   
722    723          20       RL         70.0     8120   Pave   NaN      Reg   

     LandContour Utilities  ... PoolQC  Fence MiscFeature MiscVal MoSold  \
892          Lvl    AllPub  ...    NaN  MnPrv         NaN       0      2   
1105         Lvl    AllPub  ...    NaN    NaN         NaN       0      4   
413          Lvl    AllPub  ...    NaN    NaN         NaN       0      3   
522          Lvl    AllPub  ...    NaN    NaN         NaN       0     10   
1036         HLS    AllPub  ...    NaN    NaN         NaN       0      9   
...          ...       ...  ...    ...    ...         ...     ...    ...   
479          Bnk    AllPub  ...    NaN  MnPrv         NaN       0      3   
1361         Low    AllPub  ...    NaN    NaN         NaN       0      6   
802          Lvl    AllPub  ...    NaN    NaN         NaN       0     10   
651          Lvl    AllPub  ...    NaN  MnPrv         NaN       0     10   
722          Lvl    AllPub  ...    NaN    NaN         NaN       0      7   

     YrSold SaleType  SaleCondition  SalePrice  prediction_label  
892    2006       WD         Normal     154500     136206.867118  
1105   2010       WD         Normal     325000     314469.136928  
413    2010       WD         Normal     115000     100676.612294  
522    2006       WD         Normal     159000     140646.260870  
1036   2009       WD         Normal     315500     323154.453005  
...     ...      ...            ...        ...               ...  
479    2007       WD         Alloca      89471     116587.127594  
1361   2009       WD         Normal     260000     335110.019732  
802    2008       WD         Normal     189000     221018.128964  
651    2009       WD         Normal     108000     134104.698941  
722    2009       WD         Normal     124500     125444.029402  

[292 rows x 82 columns]

In [23]:
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

# Reemplaza 'SalePrice' con el nombre real de la columna de la variable objetivo
target_column = 'SalePrice'

# Reemplaza 'Label' con el nombre real de la columna de predicciones generada por PyCaret
predictions_column = 'prediction_label'

# Calcula las métricas de rendimiento
r2 = r2_score(test_data[target_column], predictions[predictions_column])
rmse = mean_squared_error(test_data[target_column], predictions[predictions_column], squared=False)
mse = mean_squared_error(test_data[target_column], predictions[predictions_column])
mape = mean_absolute_error(test_data[target_column], predictions[predictions_column]) / test_data[target_column].mean() * 100

# Muestra las métricas
print(f'R2: {r2}\nRMSE: {rmse}\nMSE: {mse}\nMAPE: {mape}%')


R2: 0.8217839540807041
RMSE: 36972.627815128835
MSE: 1366975207.5560386
MAPE: 12.115435413007685%


#El modelo tiene un buen ajuste y explica una proporción significativa de la variabilidad en los precios de las casas. Las predicciones, en términos de RMSE, tienen una desviación promedio de alrededor de 36,972 unidades. El MAPE del 12.12% sugiere que las predicciones tienen un error moderado en términos porcentuales. Estas métricas indican un rendimiento general sólido del modelo.





